### PR TITLE
[Connectors API] Implement connector status update action

### PR DESCRIPTION
**Pull request by jedrazb**
- Implement update status endpoint
```
_connector/{connector_id}/_status
{
  "status": ....
}
```

- Use `ConnectorStateMachine` to validate state transition, throw 400 if state transition is invalid
- Add unit tests, yaml tests, bwc tests
- Tested manually

### CODE REVIEW

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/340_connector_update_status.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+
+---
+"Update Connector Status - 400 status code when invalid status transition":
+  - do:
+      catch: "bad_request"
+      connector.update_status:
+        connector_id: test-connector
+        body:
+          status: connected # created -> connected transition is invalid
```
Comment (navarone-feekery) [inline x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/340_connector_update_status.yml:92]: Nit: should we include a check that the expected error is output?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorInvalidStatusTransitionException.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+     * @param current The current state of the connector.
+     * @param next The attempted next state of the connector.
+     */
+    public ConnectorInvalidStatusTransitionException(ConnectorStatus current, ConnectorStatus next) {
+        super(
+            "Invalid transition attempt from ["
+                + current
+                + "] to ["
+                + next
+                + "]. Such a status transition is not supported by the Connectors Protocol."
```
Comment (navarone-feekery) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorInvalidStatusTransitionException.java:24]: Nit: unnecessary plural (we just call it Connector Protocol)

```suggestion
                + "]. Such a status transition is not supported by the Connector Protocol."
```

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorStateMachine.java [hunk 3]
```diff
@@ -39,6 +39,16 @@ public class ConnectorStateMachine {
...
+    }
+
+    public static void assertValidStateTransition(ConnectorStatus current, ConnectorStatus next)
+        throws ConnectorInvalidStatusTransitionException {
+        if (isValidTransition(current, next)) return;
+        throw new ConnectorInvalidStatusTransitionException(current, next);
+    }
+
+    public static Set<ConnectorStatus> validNextStates(ConnectorStatus current) {
+        return VALID_TRANSITIONS.getOrDefault(current, Collections.emptySet());
```
Comment (navarone-feekery) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorStateMachine.java:52]: Question (somewhat unrelated to this exact PR): why is `error` not a valid state following `needs_configuration`?

**Review by navarone-feekery — APPROVED** [review]
Looks good! Just left some minor comments

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorIndexServiceTests.java [hunk 4]
```diff
@@ -295,6 +296,24 @@ public void testUpdateConnectorNative() throws Exception {
         assertThat(isNative, equalTo(indexedConnector.isNative()));
     }
 
+    public void testUpdateConnectorStatus() throws Exception {
```
Comment (wangch079) [inline x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorIndexServiceTests.java:299]: Do you want to add a test case where an invalid new status is passed

File: x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/340_connector_update_status.yml (hunk 1 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/340_connector_update_status.yml:92]: Good point, I will look into it :)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorInvalidStatusTransitionException.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.application.connector;
+
+public class ConnectorInvalidStatusTransitionException extends Exception {
```
[the remark this answers is not in the document]
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorInvalidStatusTransitionException.java:10]: Yeah I'm planning to implement it for sync job as well, i wanted to leave the flexibility to adapt the error message for sync job errors.

File: x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorIndexServiceTests.java (hunk 4 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/connector/ConnectorIndexServiceTests.java:299]: Good point, I will add the unit test :)

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorStateMachine.java (hunk 3 above)
Comment (jedrazb) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorStateMachine.java:52]: I couldn't find an exact reference in the current logic, but I agree that it makes sense. I added this as a valid transition.

**Review by navarone-feekery — APPROVED** [review]
LGTM!

**Pull request merged**